client/dns_client.mli

```diff
@@ -1,34 +1,148 @@
+
+  val getaddrinfo : t -> ?nameserver:T.ns_addr -> 'response Dns.Rr_map.key ->
+    'a Domain_name.t -> ('response, [> `Msg of string ]) result T.io
+  (** [getaddrinfo nameserver query_type name] is the [query_type]-dependent
+      response from [nameserver] regarding [name], or an [Error _] message.
+      See {!Dns_client.query_state} for more information about the
+      result types.
+  *)
+
+  val gethostbyname : t -> ?nameserver:T.ns_addr -> [ `host ] Domain_name.t ->
+    (Ipaddr.V4.t, [> `Msg of string ]) result T.io
+    (** [gethostbyname state ~nameserver domain] is the IPv4 address of [domain]
+        resolved via the [state] and [nameserver] specified.
+        If the query fails, or if the [domain] does not have any IPv4 addresses,
+        an [Error _] message is returned.
+        Any extraneous IPv4 addresses are ignored.
+        For an example of using this API, see [unix/ohost.ml]
+        in the distribution of this package.
+    *)
+
+  val gethostbyname6 : t -> ?nameserver:T.ns_addr -> [ `host ] Domain_name.t ->
+    (Ipaddr.V6.t, [> `Msg of string ]) result T.io
+    (** [gethostbyname6 state ~nameserver domain] is the IPv6 address of
+        [domain] resolved via the [state] and [nameserver] specified.
+
+        It is the IPv6 equivalent of {!gethostbyname}.
+    *)
+
+end
+
+module Pure : sig
+  (** The pure interface to the client part of uDns.
+
+      Various helper modules to do with side effects are available from
+      {!Dns_client_lwt}, {!Dns_client_unix} and so forth.
+  *)
+
+  type 'key query_state constraint 'key = 'a Dns.Rr_map.key
+  (** [query_state] is parameterized over the query type, so the type of
+      the representation of the answer depends on what the name server
+      was asked to provide. See {!Dns_map.k} for a list of response types.
+      The first element (the [int32]) in most of the tuples is the
+      Time-To-Live (TTL) field returned from the server, which you can use to
+      calculate when you should request fresh information in case you are writing
+      a long-running application.
+  *)
+
+  val make_query :
+    (int -> Cstruct.t) -> Dns.proto -> 'a Domain_name.t ->
+    'query_type Dns.Rr_map.key ->
+    Cstruct.t * 'query_type Dns.Rr_map.key query_state
+  (** [make_query rng protocol name query_type] is [query, query_state]
+      where [query] is the serialized DNS query to send to the name server,
+      and [query_state] is the information required to validate the response. *)
+
+  val parse_response : 'query_type Dns.Rr_map.key query_state -> Cstruct.t ->
+    [ `Msg of string | `Ok of 'query_type | `Partial ]
+  (** [parse_response query_state response] is the information contained in
+      [response] parsed using [query_state] when the query was successful, or
+      an [`Msg message] if the [response] did not match the [query_state]
+      (or if the query failed).
+
+      In a TCP usage context the [`Partial] means there are more packets.
```

Comment:
```suggestion
      In a TCP usage context the [`Partial] means there are more bytes to be read in order to parse correctly. This can happen due to short reads or if the server (or something along the route) chunks its responses into multiple individual packets. In that case you should concatenate `response` and the next received data and call this function again.
```